WWCodeTokyo/speak-her-db
# [Part 1] Start migrating Airtable requests to Netlify functions API

First part of https://github.com/WWCodeTokyo/speak-her-db/issues/15 (needs more follow-up PRs to complete the ticket)

### Proposed Changes

- [x] Started a REST API hosted on Netlify Lambda functions under `src/lambda/` (each function is an endpoint, there should be an endpoint for each resource/entity. E.g.: people, locations, topics, languages)
    - [x] For this PR, I created only the `locations.js` function/endpoint, and I'll migrate the others progressively over the next PRs
    - [x] The locations API implement the actions `fetch` (GET method) and `create` (`POST` method), following REST standard
- [x] Added `services/api.js`: An API service to redirect requests to the lambda functions hosted on Netlify. We can import and use this service throughout the app like:
```javascript
import api from '@/services/api'; 
api.getLocations(successCallback, errorCallback);
```
- [x] Migrate `Location` requests across the app from the Airtable plugin to the new API service
    - [x] Fetch locations on `views/FindSpeaker.vue`
    - [x] Fetch locations on `components/LocationInput.vue`

**TO-DO (on follow-up PRs)**
- Migrate all other requests to Airtable to the new API service
    - Fetch topics on `views/FindSpeaker.vue` and `components/TopicsInput.vue`
    - Fetch languages on `views/FindSpeaker.vue` and `components/LanguageInput.vue`
    - Fetch speakers on `views/FindSpeaker.vue`
    - Create speaker on `components/NominateForm.vue`
- Add the caching logic to the API service for locations, topics and languages (I'll open a ticket for this cause the current caching logic is also not working, so this is basically a improvement+bug fix on top of the current implementation) 

### Testing Plan

- Check [this deploy preview](https://deploy-preview-143--speak-her-db.netlify.app/)
- When the site is loading, open the network inspector and click on the `locations` request
- Request headers now show the endpoint is `/.netlify/functions/locations` and there's no authorization key on the request headers 🔒 
![Screen Shot 2020-09-19 at 3 34 08 AM](https://user-images.githubusercontent.com/1096046/93633594-297f4b00-fa2a-11ea-98f3-ecc6ea480a56.png)
- The request response is still the same and the prefectures on the site UI show normally

Local environment:
- Pull this branch locally, run `yarn install` to install the new dependencies
- If you connection to Airtable is already setup locally as per README instructions, just run the app and follow the same steps above to see it working locally
